### ISSUE TITLE
Type inference for intersection/union templated types potentially broken when moving from `3.7.0` to `3.8.1`

### ISSUE THREAD

**Issue by Ocramius** [vimeo/psalm#2537]
It seems like psalm can no longer infer the generic types that are union/intersection based: https://travis-ci.org/Ocramius/ProxyManager/jobs/631492330#L876-L885

```
ERROR: ArgumentTypeCoercion - tests/ProxyManagerTest/Functional/AccessInterceptorScopeLocalizerFunctionalTest.php:68:13 - Argument 2 of ProxyManager\Proxy\AccessInterceptorInterface::setMethodPrefixInterceptor expects Closure(object=, object=, string=, array<string, mixed>=, bool=):mixed|null, parent type Closure(ProxyManager\Proxy\AccessInterceptorInterface, object, string, array<string, mixed>, bool):void provided

            static function (
                AccessInterceptorInterface $proxy,
                object $instance,
                string $method,
                array $params,
                bool & $returnEarly
            ) use ($listener) : void {
                $listener->__invoke($proxy, $instance, $method, $params, $returnEarly);
            }
```

In this example, the type of this callable (specifically the first parameter) is expected to be something like `AccessInterceptorInterface&TheOriginalType`, but the inference detected `object` instead

Related: https://github.com/Ocramius/ProxyManager/pull/544

**Comment by Ocramius** [comment]
@psalm-github-bot sadly not, or at least not now

**Comment by muglug** [comment]
Mind adding that to https://github.com/vimeo/psalm/blob/master/bin/test-with-real-projects.sh and I'll address it momentarily

**Comment by muglug** [comment]
Broken by b78f273ccf635fdac5b21820a022d215a5cb65a3

**Comment by muglug** [comment]
I've fixed most of the issues in d175abe0f0ee15df185846a02c8de93387619b48, the rest are caused I think by a failure to understand this: https://psalm.dev/r/f74a7b6237

**Comment by Ocramius** [comment]
I've added ProxyManager to the improvements that I had done (and forgotten, sigh) in https://github.com/vimeo/psalm/pull/2140
